src/org/thoughtcrime/securesms/ConversationFragment.java

```diff
@@ -82,688 +83,729 @@
+        ClipboardManager clipboard = (ClipboardManager) getActivity().getSystemService(Context.CLIPBOARD_SERVICE);
+        boolean first = true;
+
+        for (MessageRecord messageRecord : messageList) {
+            String body = messageRecord.getDisplayBody().toString();
+
+            if (body != null) {
+                if (!first) bodyBuilder.append('\n');
+                bodyBuilder.append(body);
+                first = false;
+            }
+        }
+
+        String result = bodyBuilder.toString();
+
+        if (!TextUtils.isEmpty(result)) {
+            clipboard.setText(result);
+        }
     }
-  }
 
-  public void showToast(String outputMessage) {
-    Toast toast=Toast.makeText(getContext(), outputMessage, Toast.LENGTH_SHORT);
-    toast.show();
-  }
+    public void handlePinOrUnpinMessage(final MessageRecord message, boolean pin,
+                                        PinnedMessageHandler handler) {
+        PinnedMessageHandler pinHandler;
+        MessagingDatabase databaseToQuery;
+        String outputMessage;
+        boolean result;
+
+        pinHandler = handler;
+        databaseToQuery = pinHandler.getAppropriateDatabase(message);
+
+        if (pin) {
+            // Blocking Video and Audio pinning Temporarily
+            if (message.isMms() && ((MmsMessageRecord) message).getSlideDeck().getSlides().size() != 0) {
+                MediaMmsMessageRecord mediaMessage = (MediaMmsMessageRecord) message;
+                if (!mediaMessage.getSlideDeck().getThumbnailSlide().getContentType().contains("image")) {
+                    showToast("You can only pin image type mms!");
+                    return;
+                }
+            }
+            result = pinHandler.handlePinMessage(message, databaseToQuery);
 
-  public interface ConversationFragmentListener {
-    void setThreadId(long threadId);
-  }
+            if (result) {
+                outputMessage = getString(R.string.ConversationFragment_pin_new);
+            } else {
+                outputMessage = getString(R.string.ConversationFragment_pin_already_pinned);
+            }
 
-  private class ConversationScrollListener extends OnScrollListener {
+        } else {
+            result = pinHandler.handleUnpinMessage(message, databaseToQuery);
 
-    private final Animation              scrollButtonInAnimation;
-    private final Animation              scrollButtonOutAnimation;
-    private final ConversationDateHeader conversationDateHeader;
+            if (result) {
+                outputMessage = getString(R.string.ConversationFragment_unpin_new);
+            } else {
+                outputMessage = getString(R.string.ConversationFragment_unpin_already_unpinned);
+            }
+        }
 
-    private boolean wasAtBottom           = true;
-    private boolean wasAtZoomScrollHeight = false;
-    private long    lastPositionId        = -1;
+        this.showToast(outputMessage);
+        searchHandler.addSearchedResult(0, message);
```

Comment:
searchHandler.addSearchedResult(0, message);
@g-harel @Dan-Yoo @shelbaz I think this was added for testing, I dont know why would you add a message for search when you are handling pin/unpin

please delete if it is left from @Dan-Yoo commenting/uncommenting the pin/unpin handle